caas/kubernetes/provider/teardown.go

```diff
@@ -0,0 +1,263 @@
+			return err
+		},
+	)
+}
+
+func (k *kubernetesClient) deleteClusterScopeCustomResourcesModelTeardown(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+) {
+	getLabels := func(crd apiextensionsv1beta1.CustomResourceDefinition) map[string]string {
+		if !isCRDNameSpaced(crd.Spec.Scope) {
+			// We only delete cluster scope CRs here, namespaced CRs are deleted by namespace destroy process.
+			return labels
+		}
+		return nil
+	}
+	ensureResourcesDeletedFunc(ctx, labels, clk, wg, errChan,
+		func(labels map[string]string) error {
+			return k.deleteCustomResources(getLabels)
+		},
+		func(labels map[string]string) error {
+			_, err := k.listCustomResources(getLabels)
+			return err
+		},
+	)
+}
+
+func (k *kubernetesClient) deleteCustomResourceDefinitionsModelTeardown(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+) {
+	ensureResourcesDeletedFunc(ctx, labels, clk, wg, errChan,
+		k.deleteCustomResourceDefinitions, func(labels map[string]string) error {
+			_, err := k.listCustomResourceDefinitions(labels)
+			return err
+		},
+	)
+}
+
+func (k *kubernetesClient) deleteMutatingWebhookConfigurationsModelTeardown(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+) {
+	ensureResourcesDeletedFunc(ctx, labels, clk, wg, errChan,
+		k.deleteMutatingWebhookConfigurations, func(labels map[string]string) error {
+			_, err := k.listMutatingWebhookConfigurations(labels)
+			return err
+		},
+	)
+}
+
+func (k *kubernetesClient) deleteValidatingWebhookConfigurationsModelTeardown(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+) {
+	ensureResourcesDeletedFunc(ctx, labels, clk, wg, errChan,
+		k.deleteValidatingWebhookConfigurations, func(labels map[string]string) error {
+			_, err := k.listValidatingWebhookConfigurations(labels)
+			return err
+		},
+	)
+}
+func (k *kubernetesClient) deleteStorageClassesModelTeardown(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+) {
+	ensureResourcesDeletedFunc(ctx, labels, clk, wg, errChan,
+		k.deleteStorageClasses, func(labels map[string]string) error {
+			_, err := k.listStorageClasses(labels)
+			return err
+		},
+	)
+}
+
+type deleterChecker func(map[string]string) error
+
+func ensureResourcesDeletedFunc(
+	ctx context.Context,
+	labels map[string]string,
+	clk jujuclock.Clock,
+	wg *sync.WaitGroup,
+	errChan chan<- error,
+	deleter, checker deleterChecker,
+) {
+	defer wg.Done()
+
+	var err error
+	defer func() {
+		if err != nil {
+			select {
+			case errChan <- err:
+			default:
+			}
+		}
+	}()
+
+	if err = deleter(labels); err != nil {
+		if errors.IsNotFound(err) {
+			err = nil
+		}
+		return
+	}
+
+	interval := 1 * time.Second
```

Comment:
not sure if this is a const somewhere(ie we use a similar polling pattern elsewhere and we can use a const to define the polling interval)

Reply:
Also 1 * time.Second == time.Second